### ISSUE TITLE
Vimeo stopped playing videos. 

### ISSUE THREAD

**Issue by greatschism** [newtaDev/pod_player#182]
`===== VIMEO API ERROR: type 'Null' is not a subtype of type 'List<dynamic>' in type cast ==========`

```
 jsonDecode(response.body)['request']['files']
          ['progressive'] 
```

"progressive" is no longer in the array, but "dash" and "his" are in the array. Found out it stopped working at 130 PM. I am waiting to hear back from vimeo. Any thoughts?

**Comment by Danzzel** [comment]
Yeah, all video are broken

**Comment by AhsanFarooq0017** [comment]
> `===== VIMEO API ERROR: type 'Null' is not a subtype of type 'List<dynamic>' in type cast ==========`
> 
> I am not sure what the problem is here, but it starts suddenly.

Hey did you find any solution?

**Comment by greatschism** [comment]
The code below works in a tester app, but starts to play then crashes my app.  I email Vimeo support to ask them about the key, and they have been less than helpful. 

The **progressive** key is missing, and the HLS and DASH are only there.  This line is null. 

`
jsonDecode(response.body)['request']['files']['progressive'] 
`

`  static Future<List<VideoQalityUrls>?> getVimeoVideoQualityUrls(
    String videoId,
    String? hash,
  ) async {
    try {
      final response = await _makeRequestHash(videoId, hash);
      var jsonData = <dynamic>[];
      var progressiveUrls = <VideoQalityUrls>[];
      // print(response.body);
      // print the key value pair of the response
      // is jsonDecode(response.body is a map or a list or something else
      // if it is a map, then proceed, else fail

      if (jsonDecode(response.body) is Map) {
        print("response.body is a map");
      } else {
        print("response.body is not a map");
      }

      if (jsonDecode(response.body)['request']['files']['progressive'] != null) {
        jsonData = jsonDecode(response.body)['request']['files']['progressive'] as List<dynamic>;
        print(jsonData);
        progressiveUrls = List.generate(
          jsonData.length,
          (index) => VideoQalityUrls(
            quality: int.parse(
              (jsonData[index]['quality'] as String?)?.split('p').first ?? '0',
            ),
            url: jsonData[index]['url'] as String,
          ),
        );
      }
      if (progressiveUrls.isEmpty && jsonDecode(response.body)['request']['files']['hls'] != null) {
        final jsonRes = jsonDecode(response.body)['request']['files']['hls']['cdns'];

        for (final element in (jsonRes as Map).entries.toList()) {
          progressiveUrls.add(
            VideoQalityUrls(
              quality: 720,
              url: element.value['url'] as String,
            ),
          );
          break;
        }
      }

      if (progressiveUrls.isEmpty && jsonDecode(response.body)['request']['files']['dash'] != null) {
        jsonData = jsonDecode(response.body)['request']['files']['dash']['cdns'] as List<dynamic>;

        progressiveUrls = List.generate(
          jsonData.length,
          (index) => VideoQalityUrls(
            quality: int.parse(
              (jsonData[index]['quality'] as String?)?.split('p').first ?? '0',
            ),
            url: jsonData[index]['url'] as String,
          ),
        );
      }
      return progressiveUrls;
    } catch (error) {
      if (error.toString().contains('XMLHttpRequest')) {
        log(
          podErrorString(
            '(INFO) To play vimeo video in WEB, Please enable CORS in your browser',
          ),
        );
      }
      debugPrint('===== 1VIMEO API ERROR: $error ==========');
      rethrow;
    }
  }`

**Comment by chatali96** [comment]
Let us know when the VIMEO API ERROR gets resolved.

**Comment by mateusfmfm** [comment]
Thx guys, I had the same problem

**Comment by newtaDev** [comment]
fixed this issue in `pod_player: ^0.2.2`
please check and let me know

**Comment by AhsanFarooq0017** [comment]
> fixed this issue in `pod_player: ^0.2.2` please check and let me know

I can not update the dependency because my app is old can you tell me any work around?

**Comment by edwardnagy** [comment]
@AhsanFarooq0017 You could clone the compatible version and apply the fix locally.

**Comment by Danzzel** [comment]
> fixed this issue in `pod_player: ^0.2.2` please check and let me know

I needed to upgrade a lot of packages but now it works fine. Thx a lot

**Comment by AhsanFarooq0017** [comment]
> @AhsanFarooq0017 You could clone the compatible version and apply the fix locally.

can you please share the complete solution?

**Comment by aaronm67** [comment]
Please do not use the /config URL [here](https://github.com/newtaDev/pod_player/blob/master/lib/src/utils/video_apis.dart#L18-L22), this is an internal API that changes without notice, so this library will break again in the future, and access to this URL will likely be disallowed in the future. Vimeo has a REST API that should be used, and has all of the data you need https://developer.vimeo.com/api/guides/start

**Comment by HasithaAcc** [comment]
I have updated the plugin with out an error but after updating to 0.2.2 video sound is not coming

this happens when we use vemio id with the hash

**Comment by greatschism** [comment]
> Please do not use the /config URL [here](https://github.com/newtaDev/pod_player/blob/master/lib/src/utils/video_apis.dart#L18-L22), this is an internal API that changes without notice, so this library will break again in the future, and access to this URL will likely be disallowed in the future. Vimeo has a REST API that should be used, and has all of the data you need https://developer.vimeo.com/api/guides/start

How do you return a player that does not need auth and does not expire?

**Comment by aaronm67** [comment]
> How do you return a player that does not need auth and does not expire?

You don't, this is intentional. You should use the REST API to get video links, which requires registering an app on https://developer.vimeo.com.

Your users can register an app on https://developer.vimeo.com and configure this player with their access token.  There are many different access tokens you can use, you can generate a "public" access key that doesn't have write permission, but we still require authentication for rate limiting purposes. Details of different types of access tokens are available here: https://developer.vimeo.com/api/guides/start#generate-access-token

If you don't wish to use the REST API, there is an OEmbed API that doesn't require authentication, https://developer.vimeo.com/api/oembed. This is only available for the Vimeo Player, not direct video links, so you'd need to embed an iFrame to support this.

Fully anonymous access to direct video files is not supported. The config URL you're currently using is an internal API that will change without notice, and will not be permanently available.

**Comment by robertjones26** [comment]
I cam across this and had to add the header with token. I searched for a couple of days and I think something changed on Vimeos side.

header = {
'Authorization': 'bearer {token}',
'Accept': 'application/vnd.vimeo.*+json;version=3.4',
};

**Comment by AAber** [comment]
I had this issue with the /config API and once I moved to the official token API the progressive issue was resolved.
If your app presents videos from multiple Vimeo accounts you will need to connect the Vimeo accounts or use multiple tokens. 
If you can't connect the Vimeo accounts make sure your app has a matching token for each Vimeo account you use.

**Comment by panalgin** [comment]
It seems like we are no longer be able to retrieve video file links without a Pro Plan on Vimeo, with their official REST API

![image](https://github.com/newtaDev/pod_player/assets/12881878/4f63fde5-a2e5-4900-8f1a-3624ea44fa73)

**Comment by hgndgn** [comment]
> I had this issue with the /config API and once I moved to the official token API the progressive issue was resolved. If your app presents videos from multiple Vimeo accounts you will need to connect the Vimeo accounts or use multiple tokens. If you can't connect the Vimeo accounts make sure your app has a matching token for each Vimeo account you use.

How are you playing the videos?

**Comment by AAber** [comment]
I'm using https://pub.dev/packages/pod_player in my Flutter app.

**Comment by hgndgn** [comment]
> I'm using https://pub.dev/packages/pod_player in my Flutter app.

Yeah me too, but the playing the video using the id and the hash does not work anymore, therefore I asked. I couldn't find any API-endpoint, where I can pass the hash of the video.

**Comment by hgndgn** [comment]
> > I'm using https://pub.dev/packages/pod_player in my Flutter app.
> 
> @AAber Yeah me too, but playing the video using the id and the hash does not work anymore, therefore I asked. I couldn't find any API-endpoint, where I can pass the hash of the video.

**Comment by hgndgn** [comment]
> I'm using https://pub.dev/packages/pod_player in my Flutter app.

@AAber  Yeah me too, but playing the video using the id and the hash does not work anymore, therefore I asked. I couldn't find any API-endpoint, where I can pass the hash of the video.

**Comment by AAber** [comment]
Use this snip to play the URL you get from the API:

```
  void initState() {
    controller = PodPlayerController(
        playVideoFrom: PlayVideoFrom.vimeo(widget.vimeoID),
        podPlayerConfig: PodPlayerConfig(
            videoQualityPriority: [360, 720, 1080], forcedVideoFocus: true))
      ..initialise();
    if (widget.videoUrls != null) {
      if (widget.videoUrls!.length != 0) {
        print('init controller');
        controller = PodPlayerController(
            playVideoFrom: PlayVideoFrom.networkQualityUrls(
              videoUrls: widget.videoUrls!,
            ),
            podPlayerConfig: PodPlayerConfig(
                videoQualityPriority: [360, 720, 1080], forcedVideoFocus: true))
          ..initialise();
      }
      super.initState();
    }
  }
```

**Comment by sanadalghnemi** [comment]
I'am facing another error here in vimeo with the access token, 
```
E/flutter ( 3580): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Exception: videoQuality cannot be empty
```

**Comment by ParanjayP-HealTech** [comment]
I have a same problem, can anyone look into this 
[Vimeo API suddenly stopped returning progressive field in video response - Issue with Flutter Android app integration](https://stackoverflow.com/questions/79429108/vimeo-api-suddenly-stopped-returning-progressive-field-in-video-response-issue)
